### DISCUSSION TITLE
How to cast as `const` with JSDoc (equivalent to `as const` in TS)

### DISCUSSION THREAD

**Discussion by voxpelli** [voxpelli/types-in-js#19] [General]
I didn't find a topic on this in here so I'm adding it.

TypeScript 3.4 introduced the concept of [_"const assertions"_](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions)  through https://github.com/Microsoft/TypeScript/pull/29510

Turns out that we can actually make use of this in the very same way as we do type casting/type assertions in JSDoc (as showed in https://github.com/voxpelli/types-in-js/discussions/18) and just like in TS simply use `const` instead of a type (as [documented here](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts)):

```javascript
// type is number[]
const arr1 = [1, 2, 3]

// type is readonly [1, 2, 3]
const arr2 = /** @type {const} */ ([1, 2, 3])
```

Example taken from @AustinGil here: https://github.com/microsoft/TypeScript/issues/30445#issuecomment-1030284467

This also works for eg. objects:

```javascript
// type is { foo: string; abc: number; }
const obj1 = { foo: 'bar', abc: 123 }

// type is { readonly foo: "bar"; readonly abc: 123; }
const obj2 = /** @type {const} */ ({ foo: 'bar', abc: 123 })
```

This can help with eg. [literal inference](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-inference) and in the [creation of enum-style definitions](https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums).

Previously I at times felt the need to separately define the type to get a similar type to work with:

```javascript
/**
 * @typedef Obj2
 * @property {'bar'} foo
 * @property {123} abc
 */

/** @type {Obj2} */
const obj2 = ({ foo: 'bar', abc: 123 })
```

Simply doing `const obj2 = /** @type {const} */ ({ foo: 'bar', abc: 123 })` is much easier and more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

**Comment by tzkmx** [comment]
How could we take advantage of the already declared `as const`-like object, to build an Union Type from the values of the object when it is used as a more flexible _enum_?

I refer to [this pattern](https://www.typescriptlang.org/play?noImplicitAny=false#code/MYewdgzgLgBAggVygCwLIFMUgCYwLwwDeAUDDAEICGEAlsAFwwDk5cAygJIDCTANKTADyiFAEZGTYUmSi+AqSgBMEhckVNiAXxjUYoSFADcxYlACeAB3TxpGLLgLmrIAGY2Ud5DgDaAa3RmrjBO6EEiaJhe2AC6xsT60DAATujA6DQAbugOzAASlADuvkw6EO4R9nEuCGDAUDTgOtLoYPXAlFDoABQhjOGeOACURALUEOhJUBwQAGqUADY02P2ROD2W6IPGZAIJIPPoAHTzIADm61ZbWibVtfWNYxNTswtLK-ZdGQsIm4yPk2UvvMfjAaGV3lERmQaG4ugBCQQAIwAVqkoIcgT8IF0IUNDjRasDsOhsZjNsMSGQyCgkiACjAwOh6QAVDYAUSStKSAk013i4ESKQAjvgiDAhT8kmZGIQmigJFRaMAmNpNHEErBKNJWVZRcLDhKJmZDlqUHFTcgWm0Ot1cdhDoq6FcLVa6DaupJwrIrkA):
```
const AuthMethod = {
  Basic: 'BASIC',
  OAuth1: 'OAuth1',
  OAuth2: 'OAuth2'
} as const;

type AuthMethod = typeof AuthMethod[keyof typeof AuthMethod];

function authenticate(method: AuthMethod) {}
```

**Reply by tzkmx** [reply]
I've tried to use simply the dynamic union type declaration but it's losing its values, it correctly validates only a literal union type.

```@jsdoc
/** @typedef { typeof AuthMethod[keyof typeof AuthMethod] } AuthMethodLosy */
/** @typedef { 'Basic' | 'OAuth1' | 'OAuth2' } AuthMethod */

authenticateFromUnion('Basic'); // works
authenticateFromUnion(/** @type AuthMethod */(AuthMethod.Basic)); // now I need to cast the value with `as AuthMethod` equivalent :(
try {
  authenticateFromUnion('hawk');
} catch (/** @type {*} */e) {
  console.warn(`catched correctly ${e.message}`)
}

authenticateFromTypeKeyof('hawk'); // it should not be accepted
authenticateFromTypeKeyof(AuthMethod.Basic); // accepted, but this form accepts any string

/**
 * @param {AuthMethodLosy} method
 */
function authenticateFromTypeKeyof(method) {
  console.log(method);
}

/**
* @param {AuthMethod} type
*/
function authenticateFromUnion(type) {
  assertIsValidAuthMethod(type);
  
  console.log(type);
}

/**
 * @param {?} value
 * @returns {asserts value is AuthMethod}
 */
function assertIsValidAuthMethod(value) {
  if (!Object.values(AuthMethod).includes(value)) {
    throw new TypeError(`Invalid authMethod: ${value}`)
  }
}
```
I leave you the [TS Playground I tried with](
https://www.typescriptlang.org/play?noImplicitAny=false&filetype=js#code/PTAEAEBcGcFoGMAWBTeBrAUB+B7AdtJKAIICukiAsshTgCagC8owAVKxJAJ4AOyoAb1wFIAX1CswAjKFAAhAIbQAlvABcoAOSKV8TaAA0M0AHkyFAIwbNZ8oguajs2xQBM1l4leaMogNxYbBxQvMh0yABmgqDcfDhR5lQ0iPQA2mjIXPExodmJ1LR0ALqg4vnJ9AAyONBcEsAYQZyh4VECWjqq+gA+Wp4OoL02id6lJHYFKQySWMKEoABOyACOTNHLpMgLXBrtCnbWnXri-rP48-sUACqha0vLAHQbW1wPl4gBGO-IeJCqCpBkAAxBY4AC2AFU8Mp8AAKbRKLoASgC31+-0BIPBUJheFhTRCfHGFEm9HqsPKhQeRyRKJYYAwkG2gmMIE4cGUAHM8DglsY0X94ADgaDIdC4ZpEAoAO5oTQo3ygIWQJCgfHsZpEgSscSSZBIlmyOY4AA2yAe0oUCzxAANlUgwkreUt4JATXUACQCc1g5DQaAKTnIUQ2pG+LACjEi8E3PgAaUy8XhUtl8r89NAyiI0BSpBNDB5RAARvwFPB4MgeIC6F87D9BcKsWDY8gE1kIhSJhU6NTEfA6WyyxWq2EDKAi+QYohlNBQBFeWDQEPKzAl3g6oQFso8JysI12DJgjwrQpFwJKVNqrVxL7CoeGhFSHhXbil3X0cro83Qm2k7epga0hGucprmiaOCcrC-70AqoiBAeR4nmeF70OIsTIBgMyPs+fz4G+FD1lGTY4nC6GAfy-pbJAACS0AAGoKCayh0ChdCwmRASyMYxpmg84GQRx4b7qwh4QMeCynoIAD84gAG6MZsongEskCkNas4CEo0BUbO8kmpsmazqxcH1Bg2EvnhWlUbRDFMSxXaFLCembORsjKFEsIAIQmEWABWqCQA8zl+p2JLdkiDzbvA+nhNATkKfqrmyFOoLSqAeDIGlLYAKILKCCywja1F4HpzH4UkhQaF6wUhmGshwaIQA).
